zedUpload/datastore_http_test.go

```diff
@@ -268,29 +271,101 @@ func testHTTPDatastoreNegative(t *testing.T) {
 	})
 }
 
+// testHTTPDatastoreRepeat tests that the client can handle lost packets
+// during a large download stream and recover, by using the Range HTTP header.
+// The tests sets up an "unstable" proxy, which is configured to start dropping and
+// delaying packets after a certain number of bytes have been transferred.
+//
+// The test works as follows:
+//
+//		1- create a large input file with random data, e.g. 100MB
+//		2- hash the input file
+//		3- create a test http server configured to serve the input file
+//		4- start the unstable proxy, pointing to the test http server as the origin,
+//		   configured to drop packets for a fixed period of time, e.g. 40 seconds, after half the file has been transferred
+//		5- download the file using the unstable proxy, which should download half the file, fail for the fixed period of time,
+//	    and then successfully download the rest
+//		6- when download completes, hash the downloaded file
+//		7- check that the hashes match
 func testHTTPDatastoreRepeat(t *testing.T) {
 	if testing.Short() {
 		t.Skip("skipping HTTP repeat test suite.")
 	} else {
 		t.Log("Running HTTP repeat test suite.")
 
-		go func() {
-			err := newUnstableProxyStart(9999, 80, "cloud-images.ubuntu.com", 1024*1024*100, 40*time.Second, 100)
-			if err != nil {
-				t.Error(err)
+		// make a random file
+		infile := httpDownloadDir + "input"
+		if _, err := os.Stat(infile); err == nil {
+			if err := os.Remove(infile); err != nil {
+				t.Fatalf("unable to remove existing file %s %v", infile, err)
+			}
+		}
+		f, err := os.Create(infile)
+		if err != nil {
+			t.Fatalf("unable to create file %s %v", infile, err)
+		}
+		defer f.Close()
+		defer os.RemoveAll(infile)
+		size := 1024 * 1024 * 100
+		bufSize := 1024 * 1024
+		randReader := io.LimitReader(rand.New(rand.NewSource(time.Now().UnixNano())), int64(size))
+		for {
+			buf := make([]byte, bufSize)
+			_, err := randReader.Read(buf)
+			if err != nil && err != io.EOF {
+				t.Fatalf("unable to read from random reader %v", err)
 			}
-		}()
+			if _, err := f.Write(buf); err != nil {
+				t.Fatalf("unable to write to file %s %v", infile, err)
+			}
+			if err == io.EOF {
+				break
+			}
+		}
+		if _, err := f.Seek(0, 0); err != nil {
+			t.Fatalf("unable to seek to beginning of file %s %v", infile, err)
+		}
+		// get the final file size
+		stat, err := os.Stat(infile)
+		if err != nil {
+			t.Fatalf("unable to stat file %s %v", infile, err)
+		}
+		insize := stat.Size()
```

Comment:
```suggestion
		inSize := stat.Size()
```
you also do `inHash`

Reply:
Changed